pkg/ccl/changefeedccl/helpers_test.go

```diff
@@ -676,6 +580,226 @@ func closeFeed(t testing.TB, f cdctest.TestFeed) {
+// of a test running as the system tenant or a secondary tenant, while still
+// having access to system interfaces for operations that require them such as
+// desctestutils.TestingGetPublicTableDescriptor
+type TestServer struct {
+	DB           *gosql.DB
+	Server       serverutils.TestTenantInterface
+	SystemDB     *gosql.DB
+	SystemServer serverutils.TestServerInterface
+	Codec        keys.SQLCodec
+	TestingKnobs base.TestingKnobs
+}
+
+func makeSystemServer(
+	t *testing.T, opts ...feedTestOption,
+) (testServer TestServer, cleanup func()) {
+	options := makeOptions(opts...)
+	return makeSystemServerWithOptions(t, options)
+}
+
+var _ = makeSystemServer // silence unused warning
+
+func makeSystemServerWithOptions(
+	t *testing.T, options feedTestOptions,
+) (testServer TestServer, cleanup func()) {
+	systemServer, systemDB, clusterCleanup := startTestFullServer(t, options)
+	return TestServer{
+			DB:           systemDB,
+			Server:       systemServer,
+			SystemDB:     systemDB,
+			SystemServer: systemServer,
+			TestingKnobs: systemServer.(*server.TestServer).Cfg.TestingKnobs,
+			Codec:        keys.SystemSQLCodec,
+		}, func() {
+			clusterCleanup()
+		}
+}
+
+func makeTenantServer(
+	t *testing.T, opts ...feedTestOption,
+) (testServer TestServer, cleanup func()) {
+	options := makeOptions(opts...)
+	return makeTenantServerWithOptions(t, options)
+}
+func makeTenantServerWithOptions(
+	t *testing.T, options feedTestOptions,
+) (testServer TestServer, cleanup func()) {
+	systemServer, systemDB, clusterCleanup := startTestFullServer(t, options)
+	tenantID, tenantServer, tenantDB, tenantCleanup := startTestTenant(t, systemServer, options)
+
+	return TestServer{
+			DB:           tenantDB,
+			Server:       tenantServer,
+			SystemDB:     systemDB,
+			SystemServer: systemServer,
+			TestingKnobs: tenantServer.(*server.TestTenant).Cfg.TestingKnobs,
+			Codec:        keys.MakeSQLCodec(tenantID),
+		}, func() {
+			tenantCleanup()
+			clusterCleanup()
+		}
+}
+
+func makeServer(t *testing.T, opts ...feedTestOption) (testServer TestServer, cleanup func()) {
+	options := makeOptions(opts...)
+	return makeServerWithOptions(t, options)
+}
+
+func makeServerWithOptions(
+	t *testing.T, options feedTestOptions,
+) (testServer TestServer, cleanup func()) {
+	if options.useTenant {
+		t.Logf("making server as secondary tenant")
+		return makeTenantServerWithOptions(t, options)
+	}
+	t.Logf("making server as system tenant")
+	return makeSystemServerWithOptions(t, options)
+}
+
+func randomSinkType(opts ...feedTestOption) string {
+	options := makeOptions(opts...)
+	return randomSinkTypeWithOptions(options)
+}
+
+func randomSinkTypeWithOptions(options feedTestOptions) string {
+	sinkWeights := map[string]int{
+		"kafka":        2, // run kafka a bit more often
+		"enterprise":   1,
+		"webhook":      1,
+		"pubsub":       1,
+		"sinkless":     0, // default to no sinkless to avoid surprises
+		"cloudstorage": 0, // requires externalIODir set
```

Comment:
Why not by default setup externalIO dir?

Reply:
Perhaps if they don't want the surprise they can .. avoid it themselves? Or randomSinkType could take an allow list.
`func randomSink(allowList ...string)`..

Reply:
See above for suggestions.  By default, all sinks; if the test cannot run on a particular sink make it explicit.